File: packages/react-native-nitro-modules/ios/core/DataArrayBuffer.swift
//
//  DataArrayBuffer.swift
//  NitroModules
//
//  Created by Marc Rousavy on 25.07.24.
//

import Foundation

public class DataBasedArrayBuffer {
  private var data: Data
  
  public init(data: Data) {
    self.data = data
  }
  
  public var size: Int {
    return data.count
  }
  
  public var bytes: UnsafeMutableRawPointer {
    var pointer: UnsafeMutableRawBufferPointer? = nil
    data.withUnsafeMutableBytes({ (p: UnsafeMutableRawBufferPointer) in
      pointer = p
    })
    return pointer!.baseAddress!
  }
}

File: packages/react-native-nitro-modules/ios/core/DataArrayBuffer.cpp
//
//  DataArrayBuffer.cpp
//  NitroModules
//
//  Created by Marc Rousavy on 25.07.24.
//

#include "DataArrayBuffer.hpp"

#if __has_include("NitroModules-Swift.h")

namespace margelo::nitro {

DataArrayBuffer::DataArrayBuffer(NitroModules::DataBasedArrayBuffer swiftPart): _swiftPart(swiftPart) { }

size_t DataArrayBuffer::size() const {
  auto swiftPart = const_cast<NitroModules::DataBasedArrayBuffer*>(&_swiftPart);
  return swiftPart->getSize();
}

uint8_t* DataArrayBuffer::data() {
  const void* bytes = _swiftPart.getBytes();
  return static_cast<uint8_t*>(const_cast<void*>(bytes));
}

} // namespace margelo::nitro

#endif
